src/ai/backend/manager/idle.py

```diff
@@ -326,6 +394,171 @@ async def update_app_streaming_status(
+            query = (
+                sa.select([keypair_resource_policies])
+                .select_from(
+                    sa.join(
+                        keypairs,
+                        keypair_resource_policies,
+                        (
+                            keypair_resource_policies.c.name
+                            == keypairs.c.resource_policy
+                        ),
+                    )
+                )
+                .where(keypairs.c.access_key == session["access_key"])
+            )
+            result = await dbconn.execute(query)
+            policy = result.first()
+            assert policy is not None
+            policy_cache[session["access_key"]] = policy
+        if policy["idle_timeout"] >= 0:
+            window_size = int(float(policy["idle_timeout"]) / interval)
+        if (window_size <= 0) or (math.isinf(window_size) and window_size > 0):
+            return True
+
+        # Get current utilization data from all containers of the session.
+        if session["cluster_size"] > 1:
+            query = (
+                sa.select([kernels.c.id])
+                .select_from(kernels)
+                .where(
+                    (kernels.c.session_id == session_id) &
+                    (kernels.c.status.in_(LIVE_STATUS))
+                )
+            )
+            result = await dbconn.execute(query)
+            rows = result.fetchall()
+            kernel_ids = [k["id"] for k in rows]
+        else:
+            kernel_ids = [session_id]
+        current_utilizations = await self.get_current_utilization(kernel_ids, occupied_slots)
+        if current_utilizations is None:
+            return True
+
+        # Update utilization time-series data.
+        not_enough_data = False
+        util_series_key = f"session.{session_id}.util_series"
+        raw_util_series = await self._redis.get(util_series_key, encoding=None)
+
+        try:
+            util_series = msgpack.unpackb(raw_util_series, use_list=True)
+        except TypeError:
+            util_series = {k: [] for k in self.resource_thresholds}
+
+        for k in util_series:
+            util_series[k].append(current_utilizations[k])
+            if len(util_series[k]) > window_size:
+                util_series[k].pop(0)
+            else:
+                not_enough_data = True
+        await self._redis.set(
+            util_series_key,
+            msgpack.packb(util_series),
+            expire=max(86400, int(self.time_window.total_seconds() * 2)),
+        )
+        if not_enough_data:
+            return True
+
+        # Check over-utilized (not to be collected) resources.
+        avg_utils = {k: sum(util_series[k]) / len(util_series[k]) for k in util_series}
+        over_utilized = {
+            k: (float(avg_utils[k]) >= float(self.resource_thresholds[k]))
+            for k in self.resource_thresholds
+            if (self.resource_thresholds[k] is not None) and (k not in unavailable_resources)
```

Comment:
We could use `for k, threshold in self.resource_thresholds.items()` for brevity and to avoid duplicate lookups.

Reply:
changed to _for k, threshold in self.resource_thresholds.items()_